src/main/scala/com/microsoft/hyperspace/index/rules/JoinIndexRule.scala

```diff
@@ -108,11 +107,12 @@ object JoinIndexRule
   private def isApplicable(l: LogicalPlan, r: LogicalPlan, condition: Expression): Boolean = {
     // The given plan is eligible if it is supported and index has not been applied.
     def isEligible(optRel: Option[FileBasedRelation]): Boolean = {
-      optRel.map(!RuleUtils.isIndexApplied(_)).getOrElse(false)
+      optRel.exists(!BaseRuleHelper.isIndexApplied(_))
     }
 
-    lazy val optLeftRel = RuleUtils.getRelation(spark, l)
-    lazy val optRightRel = RuleUtils.getRelation(spark, r)
+    lazy val ruleHelper = new BaseRuleHelper(spark)
```

Comment:
Could you let some of utility functions like getRelation in `object RuleUtils`? to avoid unnecessary instance creation

Reply:
The only method I created as outside of the class instantiation is `BaseRuleHelper.isIndexApplied` and is part of the companion object. 

All other functions do use `spark: SparkSession` inside their logic and  because of it they are now used through instantiation. 

Regarding `getRelation` that also uses `spark` inside it but we can extract that part and provide a `HyperspaceContext` parameter.